syntax = "proto3";

package api;

import "google/api/annotations.proto";
import "protoc-gen-openapiv2/options/annotations.proto";

option go_package="internal/api/pb";

option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_swagger) = {
  info: {
    title: "API"
    description: "Документация к API-сервису команды misis.tech"
    license: {
      name: "BSD 3-Clause License";
      url: "https://github.com/grpc-ecosystem/grpc-gateway/blob/main/LICENSE";
    };
  }
  security_definitions: {
    security: {
      key: "Authorization";
      value: {
        type: TYPE_API_KEY;
        in: IN_HEADER;
        name: "Authorization";
      }
    }
  }
};

service Prompter {
  rpc Extract(ExtractReq) returns (ExtractedPrompt) {
    option (google.api.http) = {
      post: "/api/v1/prompter/extract"
      body: "*"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "Извлечь метаданные из запроса"
      description: "Извлечение метаданных из запроса для передачи в предиктор"
      tags: ["prompter"]
      responses: {
        key: "200"
        value: {
          description: "Метаданные из запроса"
          examples: {
            key: "application/json"
            value: "{\"type\": 1, \"product\": \"бумага A4\", \"period\": \"18 месяцев\"}"
          }
        }
      }
      security: {
        security_requirement: {
          key: "Authorization";
          value: {}
        }
      }
    };
  }
  rpc RespondStream(StreamReq) returns (stream StreamResp) {};
}

message StreamReq {
  option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_schema) = {
    json_schema: {
      title: "Получить ответ на запрос"
      description: "Получение ответа на запрос от предиктора"
      required: ["prompt"]
    }
    example: "{\"prompt\": \"я хочу купить бумагу A4 на 18 месяцев вперед\"}"
  };
  bytes prompt = 1;
}

message StreamResp {
  option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_schema) = {
    json_schema: {
      title: "Ответ на запрос"
      description: "Ответ на запрос от предиктора по чанкам"
      required: ["chunk"]
    }
    example: "{\"chunk\": \"Предсказание: цена бумаги A4 на 18 месяцев вперед составит 1000 рублей\"}"
  };
  string chunk = 1;
}

message ExtractReq {
  option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_schema) = {
    json_schema: {
      title: "Получить метаданные из запроса"
      description: "Извлечение метаданных из запроса для передачи в предиктор"
      required: ["prompt"]
    }
    example: "{\"prompt\": \"я хочу купить бумагу A4 на 18 месяцев вперед\"}"
  };
  string prompt = 1;
}

enum QueryType {
  UNDEFINED = 0;
  PREDICTION = 1;
  STOCK = 2;
}

message ExtractedPrompt {
  option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_schema) = {
    json_schema: {
      title: "Метаданные из запроса"
      description: "Метаданные из запроса для передачи в предиктор"
      required: ["type", "product", "period"]
    }
    example: "{\"type\": 1, \"product\": \"бумага A4\", \"period\": \"18 месяцев\"}"
  };
  QueryType type = 1;
  string product = 2;
  string period = 3;
}